text/0053-enum-entities.md

```diff
@@ -0,0 +1,146 @@
+# Enumerated Entity Types
+
+## Timeline
+
+- Started: 2024-02-20
+
+## Summary
+
+Extend schemas to support declared enumerations of entity-typed values, analogous to how schemas can currently be used to enumerate a finite list of `Action`-typed values.
+
+## Basic example
+
+An enumerated entity type is declared as a normal entity type, but includes the keyword `enum` followed by the list of legal entity UIDs. Here is a simple example:
+```
+entity User;
+entity Color enum ["red", "blue", "green"];
```

Comment:
To me it felt natural to declare the entity type, first, and then to refine what UIDs it can take on, second. Migrating from TinyTodo without enums, we had
```
entity Application;
entity User in [Application];
```
and then within enums added it becomes
```
entity Application enum ["TinyTodo"];
entity User in [Application];
```
Thus users can become comfortable always writing `entity NAME` and then following it up, and not sometimes writing `enum` before the name.

Also, it looks a little weird to me with the order reversed; I want to see punctuation between the entity name and the enumerations (beyond `[`). 
```
entity enum Application ["TinyTodo"];
entity enum Color ["red", "blue", "green"];
```
You could add `::` there as in `entity enum Application :: ["TinyTodo"]` but then seems like it might be confusing too.

Reply:
I also prefer `entity enum Color`. I feel like this helps distinguish enums from other types of entities, which can have attributes & ancestors in the hierarchy.

Reply:
Also prefer `entity enum`

Reply:
I'm the odd person out, but I strongly prefer `entity Application enum [...]` over `entity enum Application [...]` for uniformity reasons.

When we write `entity` we are declaring a type, first and foremost. Then we refine that declaration with an enumeration of elements of that type. Putting the `entity enum ...` makes it seem like `enum` is the type.

Consider OCaml: You write `type id = string` or `type list = Cons of int * list | Nil`, i.e., you declare the type name first and then say what's in it. You don't prefix the type name with information about what kind of type it is. 

I'll add that other refinements/defs go after the declaration of the entity type, now. `entity Application enum ["TinyTodo"]` is similar to `entity List in [Application]` -- we are postfixing the type declaration with information about it. In the future we are leaving open the possibility of combining these things, e.g., `entity Application in [Application] enum ["TinyTodo"]` (or swapping their order) works better in terms of phrases of declaration.

Reply:
This incidentally raises another good point, that `enum` is a valid entity type, so `entity enum in [Group]` and the like are valid today -- potentially a parsing problem (or at least confusion) for the `entity enum Application` option